File: src/09.tree/57.235.go
package tree

// root 是 二叉搜索树
func LowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// 保证 p.Val < q.Val
	if p.Val > q.Val {
		tmp := p
		p = q
		q = tmp
	}

	for root != nil {
		if root.Val < p.Val {
			root = root.Right
		} else if root.Val > q.Val {
			root = root.Left
		} else {
			break
		}
	}

	return root
}

File: src/09.tree/57.235.ts
import { TreeNode } from "./01.100";

// root 是 二叉搜索树
export function lowestCommonAncestor(
    root: TreeNode | null,
    p: TreeNode | null,
    q: TreeNode | null
): TreeNode | null {
    if (root.val < p.val && root.val < q.val) {
        return lowestCommonAncestor(root.right, p, q);
    } else if (root.val > p.val && root.val > q.val) {
        return lowestCommonAncestor(root.left, p, q);
    }
    return root;
}
